tests/frontend/io/test_io_xir.py

```diff
@@ -0,0 +1,646 @@
+
+    with prog.context as q:
+        # state preparation
+        ops.Vac | q[1]
+        ops.Squeezed(0.12) | q[2]
+
+        # one mode gates
+        ops.Sgate(1) | q[0]
+        ops.Dgate(np.abs(0.54 + 0.5j), np.angle(0.54 + 0.5j)) | q[1]
+
+        # two mode gates
+        ops.S2gate(0.543, -0.12) | (q[0], q[3])
+
+        # decomposition
+        ops.Interferometer(U) | q
+
+        # measurement
+        ops.MeasureX | q[0]
+        ops.MeasureHomodyne(0.43, select=0.32) | q[2]
+        ops.MeasureHomodyne(phi=0.43, select=0.32) | q[2]
+
+    return prog
+
+
+test_xir_prog_not_compiled = """\
+Vacuum | [1];
+Squeezed(0.12, 0.0) | [2];
+Sgate(1, 0.0) | [0];
+Dgate(0.735934779718964, 0.7469555733762603) | [1];
+S2gate(0.543, -0.12) | [0, 3];
+Interferometer([[(0.219546940711-0.256534554457j), (0.611076853957+0.524178937791j), (-0.102700187435+0.474478834685j), (-0.027250232925+0.03729094623j)], [(0.451281863394+0.602582912475j), (0.456952590016+0.01230749109j), (0.131625867435-0.450417744715j), (0.035283194078-0.053244267184j)], [(0.038710094355+0.492715562066j), (-0.019212744068-0.321842852355j), (-0.240776471286+0.524432833034j), (-0.458388143039+0.329633367819j)], [(-0.156619083736+0.224568570065j), (0.109992223305-0.163750223027j), (-0.421179844245+0.183644837982j), (0.818769184612+0.068015658737j)]]) | [0, 1, 2, 3];
+MeasureHomodyne(phi: 0) | [0];
+MeasureHomodyne(phi: 0.43, select: 0.32) | [2];
+MeasureHomodyne(phi: 0.43, select: 0.32) | [2];\
+"""
+
+
+class TestSFtoXIRConversion:
+    """Tests for the io.to_xir utility function"""
+
+    def test_empty_program(self):
+        """Test that an empty program is correctly converted"""
+        # create a test program
+        sf_prog = Program(4, name="test_program")
+        xir_prog = io.to_xir(sf_prog)
+
+        assert xir_prog.serialize() == ""
+        assert xir_prog.version == "0.1.0"
+
+    def test_gate_noarg(self):
+        """Test gate with no argument converts"""
+        # create a test program
+        sf_prog = Program(1)
+
+        with sf_prog.context as q:
+            ops.Vac | q[0]
+
+        xir_prog = io.to_xir(sf_prog)
+
+        expected = [("Vacuum", [], (0,))]
+        assert [(stmt.name, stmt.params, stmt.wires) for stmt in xir_prog.statements] == expected
+
+    def test_gate_arg(self):
+        """Test gate with argument converts"""
+        # create a test program
+        sf_prog = Program(2)
+
+        with sf_prog.context as q:
+            ops.Sgate(0.54, 0.324) | q[1]
+
+        xir_prog = io.to_xir(sf_prog)
+
+        expected = [("Sgate", [0.54, 0.324], (1,))]
+        assert [(stmt.name, stmt.params, stmt.wires) for stmt in xir_prog.statements] == expected
+
+    def test_gate_kwarg(self):
+        """Test gate with keyword argument converts"""
+        # create a test program
+        sf_prog = Program(2)
+
+        with sf_prog.context as q:
+            ops.Dgate(r=np.abs(0.54 + 0.324j), phi=np.angle(0.54 + 0.324j)) | q[1]
```

Comment:
Can we try using mode 0 instead for completeness sake?
```suggestion
            ops.Dgate(r=np.abs(0.54 + 0.324j), phi=np.angle(0.54 + 0.324j)) | q[0]
```

Reply:
Good point! 👍